### PR TITLE
[Merged by Bors] - feat: compute rank of `ℂ` and `ℝ` as `ℚ`-modules

### PR DESCRIPTION
**Pull request by j-loreaux**
This adds a trivial cardinality argument that shows that when `V` is a free `K`-module where `K` is infinite and satisfies the strong rank condiiton, then the rank of `V` coincides with its cardinality. This is then used to establish that `Module.rank ℚ ℝ = continuum = Module.rank ℚ ℂ`, and therefore that `ℝ` and `ℂ` are isomorphic as vector spaces over `ℚ`.

As requested on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Hamel.20basis.20and.20reals.20isomorphic.20to.20the.20complexes.20as.20an.20a.2E.2E.2E)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 1]
```diff
@@ -935,6 +935,14 @@ theorem rank_eq_card_chooseBasisIndex : Module.rank K V = #(ChooseBasisIndex K V
   (chooseBasis K V).mk_eq_rank''.symm
 #align module.free.rank_eq_card_choose_basis_index Module.Free.rank_eq_card_chooseBasisIndex
 
+lemma rank_eq_mk_of_infinite_lt [Infinite K] (h_lt : lift.{v, u} #K < lift.{max u v, v} #V) :
```
Comment (pechersky) [inline Mathlib/LinearAlgebra/Dimension.lean:938]: The minor-est nit -- should `K` and `V` be implicit here? Won't be useful in your `rank_rat_*`, but according to guidelines, they're both present in `h_lt`.

File: Mathlib/Data/Complex/Module.lean [hunk 2]
```diff
@@ -472,3 +472,24 @@ theorem imaginaryPart_smul (z : ℂ) (a : A) : ℑ (z • a) = z.re • ℑ a +
...
+
+open Cardinal Module
+
+@[simp]
+lemma Real.rank_rat_real : Module.rank ℚ ℝ = continuum := by
+  refine (Free.rank_eq_mk_of_infinite_lt ℚ ℝ ?_).trans mk_real
+  simpa [mk_real] using aleph0_lt_continuum
+
+@[simp]
+lemma Complex.rank_rat_complex : Module.rank ℚ ℂ = continuum := by
```
Comment (pechersky) [inline Mathlib/Data/Complex/Module.lean:486]: If already in the `Real` (ditto Complex) namespace, do we need `_real` at the end?

File: Mathlib/Data/Complex/Module.lean [hunk 3]
```diff
@@ -472,3 +472,24 @@ theorem imaginaryPart_smul (z : ℂ) (a : A) : ℑ (z • a) = z.re • ℑ a +
...
+  refine (Free.rank_eq_mk_of_infinite_lt ℚ ℂ ?_).trans mk_complex
+  simpa [mk_real] using aleph0_lt_continuum
+
+/-- `ℂ` and `ℝ` are isomorphic as vector spaces over `ℚ`, or equivalently,
+as additive grops. -/
```
Comment (pechersky) [inline Mathlib/Data/Complex/Module.lean:491]:
```suggestion
as additive groups. -/
```
typo

File: Mathlib/Data/Complex/Module.lean [hunk 4]
```diff
@@ -472,3 +472,24 @@ theorem imaginaryPart_smul (z : ℂ) (a : A) : ℑ (z • a) = z.re • ℑ a +
...
+
+@[simp]
+lemma Real.rank_rat_real : Module.rank ℚ ℝ = continuum := by
+  refine (Free.rank_eq_mk_of_infinite_lt ℚ ℝ ?_).trans mk_real
+  simpa [mk_real] using aleph0_lt_continuum
+
+@[simp]
+lemma Complex.rank_rat_complex : Module.rank ℚ ℂ = continuum := by
+  refine (Free.rank_eq_mk_of_infinite_lt ℚ ℂ ?_).trans mk_complex
+  simpa [mk_real] using aleph0_lt_continuum
```
Comment (pechersky) [inline Mathlib/Data/Complex/Module.lean:488]: Is `mk_real` necessary here too?

**Review by pechersky — APPROVED** [review]
Thank you!

File: Mathlib/LinearAlgebra/Dimension.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/LinearAlgebra/Dimension.lean:938]: And I think the condition is usually written simply as `lift.{v} #K < lift.{u} #V` (i.e. `lift.{v u} #K < lift.{u v} #V`. This is the case, for example, in [this theorem you use](https://github.com/leanprover-community/mathlib4/blob/96f0a9003a1032e51a21b807aa9fe772112b5816/Mathlib/SetTheory/Cardinal/Ordinal.lean#L1049-L1056).

File: Mathlib/Data/Complex/Module.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Data/Complex/Module.lean:486]: Yeah, `Complex.rank_rat` or `Module.rank_rat_complex` both seem like better names. What's the lemma that `Module.rank ℚ ℝ = 2` called?

Comment (j-loreaux) [reply Mathlib/Data/Complex/Module.lean:486]: yes, you found why I named it the way I did, but I see your points.

File: Mathlib/Data/Complex/Module.lean (hunk 4 above)
Comment (j-loreaux) [reply Mathlib/Data/Complex/Module.lean:488]: nope, thanks!

File: Mathlib/Data/Complex/Module.lean [hunk 5]
```diff
@@ -485,10 +485,10 @@ lemma Real.rank_rat_real : Module.rank ℚ ℝ = continuum := by
 @[simp]
 lemma Complex.rank_rat_complex : Module.rank ℚ ℂ = continuum := by
   refine (Free.rank_eq_mk_of_infinite_lt ℚ ℂ ?_).trans mk_complex
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:487]: These proofs look like good arguments for reverting the arguments to explicit...

bors d+

Either way is fine by be

**Pull request closed without merging**